store/tikv/lock_resolver.go

```diff
@@ -578,10 +601,253 @@ func (lr *LockResolver) getTxnStatus(bo *Backoffer, txnID uint64, primary []byte
 			status.commitTS = cmdResp.CommitVersion
 			lr.saveResolved(txnID, status)
 		}
+
 		return status, nil
 	}
 }
 
+// asyncResolveData is data contributed by multiple goroutines when resolving locks using the async commit protocol. All
+// data should be protected by the mutex field.
+type asyncResolveData struct {
+	mutex sync.Mutex
+	// If any lock has been committed, then this is the commit ts. In that case, all locks should have been committed
+	// with the same commit timestamp. If no locks have been committed, then we will use the min commit ts from the
+	// primary lock. Note that a secondary lock's commit ts may or may not be the same as the primary lock's min commit
+	// ts.
+	commitTs    uint64
+	locks       map[RegionVerID][]*Lock
+	missingLock bool
+}
+
+// addLocks adds locks to data, keeping other fields up to date. startTS and commitTS are for the transaction being
+// resolved.
+//
+// In the async commit protocol when checking locks, we send a list of keys to check and get back a list of locks. There
+// will be a lock for every key which is locked. If there are fewer locks than keys, then a lock is missing because it
+// has been committed, rolled back, or was never locked.
+//
+// In this function, locks is the list of locks, and expected is the number of keys. asyncResolveData.missingLock will be
+// set to true if the lengths don't match. If the lengths do match, then the locks are added to asyncResolveData.locks
+// and will need to be resolved by the caller.
+func (data *asyncResolveData) addLocks(locks []*kvrpcpb.LockInfo, curRegionID RegionVerID, expected int, startTS uint64, commitTS uint64) error {
+	data.mutex.Lock()
+	defer data.mutex.Unlock()
+
+	// Check locks to see if any have been committed or rolled back.
+	if len(locks) < expected {
+		logutil.BgLogger().Info("addLocks lock has been committed or rolled back", zap.Uint64("commit ts", commitTS), zap.Uint64("start ts", startTS))
+		// A lock is missing - the transaction must either have been rolled back or committed.
+		if !data.missingLock {
+			// commitTS == 0 => lock has been rolled back.
+			if commitTS != 0 && commitTS < data.commitTs {
+				return errors.Errorf("commit TS must be greater or equal to min commit TS: commit ts: %v, min commit ts: %v", commitTS, data.commitTs)
+			}
+			data.commitTs = commitTS
+		}
+		data.missingLock = true
+
+		if data.commitTs != commitTS {
+			return errors.Errorf("commit TS mismatch in async commit recovery: %v and %v", data.commitTs, commitTS)
+		}
+
+		// We do not need to resolve the remaining locks because TiKV will have resolved them as appropriate.
```

Comment:
Oh, this looks like a TODO is TiKV, right?

Reply:
This is only locks in the same region, and TiKV should have done this already in response to the check secondary locks message.